### PR TITLE
refactor(all): refactor for lint tests

### LINKED ISSUE #43: lint failing
**I'm submitting a bug report**

* **Library Version:**
0.1.16

**Please tell us about your environment:**
* **Operating System:**
Windows 10

* **Node Version:**
7.7.4

* **NPM Version:**
4.1.2

* **Language:**
TypeScript 2.2.1

**Current behavior:**
lint operation errors with the following output

```
> aurelia-ux@0.1.16 lint E:\Development\Aurelia\ux
> cross-env tslint --project tsconfig.json

Warning: Got empty options

E:/Development/Aurelia/ux/src/hosts/cordova.ts[35, 9]: Identifier 'device' is never reassigned; use 'const' instead of 'let'.
E:/Development/Aurelia/ux/src/hosts/electron.ts[14, 9]: Identifier 'p' is never reassigned; use 'const' instead of 'let'.
E:/Development/Aurelia/ux/src/styles/style-controller.ts[52, 27]: Type assertion using the '<>' syntax is forbidden. Use the 'as' syntax instead.
E:/Development/Aurelia/ux/src/styles/style-controller.ts[23, 9]: Identifier 'overrideContext' is never reassigned; use 'const' instead of 'let'.
E:/Development/Aurelia/ux/src/styles/style-controller.ts[24, 9]: Identifier '$styles' is never reassigned; use 'const' instead of 'let'.
E:/Development/Aurelia/ux/src/styles/style-factory.ts[37, 38]: Parentheses are required around the parameters of an arrow function definition
E:/Development/Aurelia/ux/src/styles/style-factory.ts[61, 1]: A maximum of 1 class per file is allowed
E:/Development/Aurelia/ux/src/styles/style-factory.ts[31, 11]: Identifier 'baseStyles' is never reassigned; use 'const' instead of 'let'.
E:/Development/Aurelia/ux/src/styles/style-factory.ts[27, 9]: Identifier '$styles' is never reassigned; use 'const' instead of 'let'.
E:/Development/Aurelia/ux/src/styles/style-factory.ts[28, 9]: Identifier 'ux' is never reassigned; use 'const' instead of 'let'.
E:/Development/Aurelia/ux/src/styles/style-factory.ts[13, 40]: Don't use 'Function' as a type. Avoid using the `Function` type. Prefer a specific function type, like `() => void`.
E:/Development/Aurelia/ux/src/styles/style-factory.ts[13, 96]: Don't use 'Object' as a type. Avoid using the `Object` type. Did you mean `object`?
E:/Development/Aurelia/ux/src/styles/style-compiler.ts[34, 1]: A maximum of 1 class per file is allowed
E:/Development/Aurelia/ux/src/styles/style-compiler.ts[42, 1]: A maximum of 1 class per file is allowed
E:/Development/Aurelia/ux/src/styles/style-compiler.ts[19, 22]: Type assertion using the '<>' syntax is forbidden. Use the 'as' syntax instead.
E:/Development/Aurelia/ux/src/styles/style-compiler.ts[14, 11]: Identifier 'name' is never reassigned; use 'const' instead of 'let'.
E:/Development/Aurelia/ux/src/styles/style-compiler.ts[12, 9]: Identifier 'styles' is never reassigned; use 'const' instead of 'let'.
E:/Development/Aurelia/ux/src/styles/style-compiler.ts[13, 9]: Identifier 'transformed' is never reassigned; use 'const' instead of 'let'.
E:/Development/Aurelia/ux/src/styles/style-compiler.ts[11, 35]: Don't use 'Function' as a type. Avoid using the `Function` type. Prefer a specific function type, like `() => void`.
E:/Development/Aurelia/ux/src/styles/style-strategy.ts[112, 1]: A maximum of 1 class per file is allowed
E:/Development/Aurelia/ux/src/styles/style-strategy.ts[147, 1]: A maximum of 1 class per file is allowed
E:/Development/Aurelia/ux/src/styles/style-strategy.ts[19, 41]: Type assertion using the '<>' syntax is forbidden. Use the 'as' syntax instead.
E:/Development/Aurelia/ux/src/styles/style-strategy.ts[78, 30]: Type assertion using the '<>' syntax is forbidden. Use the 'as' syntax instead.
E:/Development/Aurelia/ux/src/styles/style-strategy.ts[93, 24]: Type assertion using the '<>' syntax is forbidden. Use the 'as' syntax instead.
E:/Development/Aurelia/ux/src/styles/style-strategy.ts[138, 24]: Type assertion using the '<>' syntax is forbidden. Use the 'as' syntax instead.
E:/Development/Aurelia/ux/src/styles/style-strategy.ts[163, 20]: Type assertion using the '<>' syntax is forbidden. Use the 'as' syntax instead.
E:/Development/Aurelia/ux/src/styles/style-strategy.ts[172, 23]: Type assertion using the '<>' syntax is forbidden. Use the 'as' syntax instead.
E:/Development/Aurelia/ux/src/styles/style-strategy.ts[42, 9]: Identifier 'quote' is never reassigned; use 'const' instead of 'let'.
E:/Development/Aurelia/ux/src/styles/style-strategy.ts[75, 11]: Identifier 'path' is never reassigned; use 'const' instead of 'let'.
E:/Development/Aurelia/ux/src/styles/style-strategy.ts[93, 13]: Identifier 'compiler' is never reassigned; use 'const' instead of 'let'.
E:/Development/Aurelia/ux/src/styles/style-strategy.ts[85, 9]: Identifier 'styleUrl' is never reassigned; use 'const' instead of 'let'.
E:/Development/Aurelia/ux/src/styles/style-strategy.ts[138, 13]: Identifier 'compiler' is never reassigned; use 'const' instead of 'let'.
E:/Development/Aurelia/ux/src/styles/style-strategy.ts[161, 9]: Identifier 'css' is never reassigned; use 'const' instead of 'let'.
E:/Development/Aurelia/ux/src/styles/style-strategy.ts[163, 9]: Identifier 'compiler' is never reassigned; use 'const' instead of 'let'.
E:/Development/Aurelia/ux/src/styles/style-strategy.ts[172, 9]: Identifier 'designType' is never reassigned; use 'const' instead of 'let'.
E:/Development/Aurelia/ux/src/styles/style-strategy.ts[34, 5]: Identifier 'cssUrlMatcher' is never reassigned; use 'const' instead of 'let'.
E:/Development/Aurelia/ux/src/styles/style-strategy.ts[13, 59]: Don't use 'Function' as a type. Avoid using the `Function` type. Prefer a specific function type, like `() => void`.
E:/Development/Aurelia/ux/src/styles/style-strategy.ts[19, 29]: Don't use 'Function' as a type. Avoid using the `Function` type. Prefer a specific function type, like `() => void`.
E:/Development/Aurelia/ux/src/styles/style-strategy.ts[73, 66]: Don't use 'Function' as a type. Avoid using the `Function` type. Prefer a specific function type, like `() => void`.
E:/Development/Aurelia/ux/src/styles/style-strategy.ts[131, 66]: Don't use 'Function' as a type. Avoid using the `Function` type. Prefer a specific function type, like `() => void`.
E:/Development/Aurelia/ux/src/styles/style-strategy.ts[160, 66]: Don't use 'Function' as a type. Avoid using the `Function` type. Prefer a specific function type, like `() => void`.
E:/Development/Aurelia/ux/src/styles/style-locator.ts[28, 8]: Type assertion using the '<>' syntax is forbidden. Use the 'as' syntax instead.
E:/Development/Aurelia/ux/src/styles/style-locator.ts[41, 10]: Type assertion using the '<>' syntax is forbidden. Use the 'as' syntax instead.
E:/Development/Aurelia/ux/src/styles/style-locator.ts[50, 20]: Type assertion using the '<>' syntax is forbidden. Use the 'as' syntax instead.
E:/Development/Aurelia/ux/src/styles/style-locator.ts[20, 11]: Identifier 'origin' is never reassigned; use 'const' instead of 'let'.
E:/Development/Aurelia/ux/src/styles/style-locator.ts[49, 9]: Identifier 'origin' is never reassigned; use 'const' instead of 'let'.
E:/Development/Aurelia/ux/src/styles/style-locator.ts[82, 9]: Identifier 'moduleId' is never reassigned; use 'const' instead of 'let'.
E:/Development/Aurelia/ux/src/styles/style-locator.ts[83, 9]: Identifier 'id' is never reassigned; use 'const' instead of 'let'.
E:/Development/Aurelia/ux/src/styles/style-engine.ts[22, 30]: Type assertion using the '<>' syntax is forbidden. Use the 'as' syntax instead.
E:/Development/Aurelia/ux/src/styles/style-engine.ts[29, 25]: Type assertion using the '<>' syntax is forbidden. Use the 'as' syntax instead.
E:/Development/Aurelia/ux/src/styles/style-engine.ts[30, 10]: Type assertion using the '<>' syntax is forbidden. Use the 'as' syntax instead.
E:/Development/Aurelia/ux/src/styles/style-engine.ts[51, 25]: Type assertion using the '<>' syntax is forbidden. Use the 'as' syntax instead.
E:/Development/Aurelia/ux/src/styles/style-engine.ts[59, 8]: Type assertion using the '<>' syntax is forbidden. Use the 'as' syntax instead.
E:/Development/Aurelia/ux/src/styles/style-engine.ts[70, 23]: Type assertion using the '<>' syntax is forbidden. Use the 'as' syntax instead.
E:/Development/Aurelia/ux/src/styles/style-engine.ts[76, 10]: Type assertion using the '<>' syntax is forbidden. Use the 'as' syntax instead.
E:/Development/Aurelia/ux/src/styles/style-engine.ts[78, 10]: Type assertion using the '<>' syntax is forbidden. Use the 'as' syntax instead.
E:/Development/Aurelia/ux/src/styles/style-engine.ts[21, 9]: Identifier 'themeKey' is never reassigned; use 'const' instead of 'let'.
E:/Development/Aurelia/ux/src/styles/style-engine.ts[22, 9]: Identifier 'currentController' is never reassigned; use 'const' instead of 'let'.
E:/Development/Aurelia/ux/src/styles/style-engine.ts[73, 11]: Identifier 'shadowDOMRoot' is never reassigned; use 'const' instead of 'let'.
E:/Development/Aurelia/ux/src/styles/style-engine.ts[86, 9]: Identifier 'root' is never reassigned; use 'const' instead of 'let'.
E:/Development/Aurelia/ux/src/styles/style-engine.ts[20, 57]: Don't use 'Object' as a type. Avoid using the `Object` type. Did you mean `object`?
E:/Development/Aurelia/ux/src/styles/style-resource.ts[34, 81]: Parentheses are required around the parameters of an arrow function definition
E:/Development/Aurelia/ux/src/styles/style-resource.ts[42, 1]: A maximum of 1 class per file is allowed
E:/Development/Aurelia/ux/src/styles/style-resource.ts[27, 26]: Type assertion using the '<>' syntax is forbidden. Use the 'as' syntax instead.
E:/Development/Aurelia/ux/src/styles/style-resource.ts[27, 9]: Identifier 'styleStrategy' is never reassigned; use 'const' instead of 'let'.
E:/Development/Aurelia/ux/src/styles/style-resource.ts[9, 27]: Don't use 'Function' as a type. Avoid using the `Function` type. Prefer a specific function type, like `() => void`.
E:/Development/Aurelia/ux/src/styles/style-resource.ts[16, 51]: Don't use 'Function' as a type. Avoid using the `Function` type. Prefer a specific function type, like `() => void`.
E:/Development/Aurelia/ux/src/styles/decorators.ts[19, 19]: Don't use 'Function' as a type. Avoid using the `Function` type. Prefer a specific function type, like `() => void`.
E:/Development/Aurelia/ux/src/styles/dynamic-styles.ts[10, 16]: Don't use 'Function' as a type. Avoid using the `Function` type. Prefer a specific function type, like `() => void`.
E:/Development/Aurelia/ux/src/ux-configuration.ts[34, 17]: Type assertion using the '<>' syntax is forbidden. Use the 'as' syntax instead.
E:/Development/Aurelia/ux/src/ux-configuration.ts[42, 17]: Type assertion using the '<>' syntax is forbidden. Use the 'as' syntax instead.
E:/Development/Aurelia/ux/src/ux-configuration.ts[34, 9]: Identifier 'proto' is never reassigned; use 'const' instead of 'let'.
E:/Development/Aurelia/ux/src/ux-configuration.ts[35, 9]: Identifier 'original' is never reassigned; use 'const' instead of 'let'.
src/aurelia-ux.ts[28, 42]: Parentheses are required around the parameters of an arrow function definition
src/aurelia-ux.ts[36, 41]: Parentheses are required around the parameters of an arrow function definition
src/aurelia-ux.ts[28, 9]: Identifier 'found' is never reassigned; use 'const' instead of 'let'.
src/index.ts[22, 12]: Type assertion using the '<>' syntax is forbidden. Use the 'as' syntax instead.
src/index.ts[22, 7]: Identifier 'ux' is never reassigned; use 'const' instead of 'let'.
E:/Development/Aurelia/ux/src/designs/design-attributes.ts[16, 11]: Identifier 'realName' is never reassigned; use 'const' instead of 'let'.
E:/Development/Aurelia/ux/src/designs/design-attributes.ts[13, 9]: Identifier 'current' is never reassigned; use 'const' instead of 'let'.
E:/Development/Aurelia/ux/src/designs/design-attributes.ts[12, 19]: Identifier 'ii' is never reassigned; use 'const' instead of 'let'.
E:/Development/Aurelia/ux/src/designs/design-attributes.ts[9, 7]: Identifier 'prefix' is never reassigned; use 'const' instead of 'let'.
E:/Development/Aurelia/ux/src/designs/design-attributes.ts[10, 7]: Identifier 'length' is never reassigned; use 'const' instead of 'let'.
E:/Development/Aurelia/ux/src/effects/element-rect.ts[78, 1]: Exceeds maximum line length of 120
E:/Development/Aurelia/ux/src/effects/element-rect.ts[81, 5]: Multiple variable declarations in the same statement are forbidden
E:/Development/Aurelia/ux/src/effects/element-rect.ts[85, 7]: variable name must be in camelcase or uppercase
E:/Development/Aurelia/ux/src/effects/element-rect.ts[86, 7]: variable name must be in camelcase or uppercase
E:/Development/Aurelia/ux/src/effects/element-rect.ts[81, 9]: Identifier 'l' is never reassigned; use 'const' instead of 'let'.
E:/Development/Aurelia/ux/src/effects/element-rect.ts[82, 7]: Identifier 't' is never reassigned; use 'const' instead of 'let'.
E:/Development/Aurelia/ux/src/effects/element-rect.ts[83, 7]: Identifier 'w' is never reassigned; use 'const' instead of 'let'.
E:/Development/Aurelia/ux/src/effects/element-rect.ts[84, 7]: Identifier 'h' is never reassigned; use 'const' instead of 'let'.
E:/Development/Aurelia/ux/src/effects/element-rect.ts[85, 7]: Identifier '_x' is never reassigned; use 'const' instead of 'let'.
E:/Development/Aurelia/ux/src/effects/element-rect.ts[86, 7]: Identifier '_y' is never reassigned; use 'const' instead of 'let'.
E:/Development/Aurelia/ux/src/effects/paper-wave.ts[5, 1]: Multiple variable declarations in the same statement are forbidden
E:/Development/Aurelia/ux/src/effects/paper-wave.ts[158, 5]: Multiple variable declarations in the same statement are forbidden
E:/Development/Aurelia/ux/src/effects/paper-wave.ts[226, 5]: Multiple variable declarations in the same statement are forbidden
E:/Development/Aurelia/ux/src/effects/paper-wave.ts[268, 5]: Multiple variable declarations in the same statement are forbidden
E:/Development/Aurelia/ux/src/effects/paper-wave.ts[5, 5]: variable name must be in camelcase or uppercase
E:/Development/Aurelia/ux/src/effects/paper-wave.ts[6, 3]: variable name must be in camelcase or uppercase
E:/Development/Aurelia/ux/src/effects/paper-wave.ts[7, 3]: variable name must be in camelcase or uppercase
E:/Development/Aurelia/ux/src/effects/paper-wave.ts[158, 9]: Identifier 'radius' is never reassigned; use 'const' instead of 'let'.
E:/Development/Aurelia/ux/src/effects/paper-wave.ts[162, 7]: Identifier 'elapsed' is never reassigned; use 'const' instead of 'let'.
E:/Development/Aurelia/ux/src/effects/paper-wave.ts[163, 7]: Identifier 'currentTime' is never reassigned; use 'const' instead of 'let'.
E:/Development/Aurelia/ux/src/effects/paper-wave.ts[164, 7]: Identifier 'actualRadius' is never reassigned; use 'const' instead of 'let'.
E:/Development/Aurelia/ux/src/effects/paper-wave.ts[226, 9]: Identifier 'translateFraction' is never reassigned; use 'const' instead of 'let'.
E:/Development/Aurelia/ux/src/effects/paper-wave.ts[268, 9]: Identifier 'scaleFactor' is never reassigned; use 'const' instead of 'let'.
E:/Development/Aurelia/ux/src/effects/paper-wave.ts[269, 7]: Identifier 'containerCenter' is never reassigned; use 'const' instead of 'let'.
E:/Development/Aurelia/ux/src/effects/paper-wave.ts[270, 7]: Identifier 'currentPos' is never reassigned; use 'const' instead of 'let'.
E:/Development/Aurelia/ux/src/effects/paper-wave.ts[271, 7]: Identifier 'deltaPos' is never reassigned; use 'const' instead of 'let'.
E:/Development/Aurelia/ux/src/effects/paper-wave.ts[301, 9]: Identifier 'containerCenter' is never reassigned; use 'const' instead of 'let'.
E:/Development/Aurelia/ux/src/effects/paper-wave.ts[5, 5]: Identifier '_window' is never reassigned; use 'const' instead of 'let'.
E:/Development/Aurelia/ux/src/effects/paper-wave.ts[6, 3]: Identifier '_doc' is never reassigned; use 'const' instead of 'let'.
E:/Development/Aurelia/ux/src/effects/paper-wave.ts[7, 3]: Identifier '_now' is never reassigned; use 'const' instead of 'let'.
E:/Development/Aurelia/ux/src/effects/paper-ripple.ts[109, 28]: Parentheses are required around the parameters of an arrow function definition
E:/Development/Aurelia/ux/src/effects/paper-ripple.ts[165, 24]: Parentheses are required around the parameters of an arrow function definition
E:/Development/Aurelia/ux/src/effects/paper-ripple.ts[113, 9]: variable name must be in camelcase or uppercase
E:/Development/Aurelia/ux/src/effects/paper-ripple.ts[120, 9]: Type assertion using the '<>' syntax is forbidden. Use the 'as' syntax instead.
E:/Development/Aurelia/ux/src/effects/paper-ripple.ts[127, 9]: Type assertion using the '<>' syntax is forbidden. Use the 'as' syntax instead.
E:/Development/Aurelia/ux/src/effects/paper-ripple.ts[113, 9]: Identifier '_doc' is never reassigned; use 'const' instead of 'let'.
E:/Development/Aurelia/ux/src/effects/paper-ripple.ts[140, 9]: Identifier 'wave' is never reassigned; use 'const' instead of 'let'.
E:/Development/Aurelia/ux/src/effects/paper-ripple.ts[154, 9]: Identifier 'wave' is never reassigned; use 'const' instead of 'let'.
E:/Development/Aurelia/ux/src/effects/paper-ripple.ts[176, 9]: Identifier 'waveIndex' is never reassigned; use 'const' instead of 'let'.
E:/Development/Aurelia/ux/src/effects/paper-ripple.ts[193, 11]: Identifier 'wave' is never reassigned; use 'const' instead of 'let'.
E:/Development/Aurelia/ux/src/effects/paper-ripple.ts[192, 21]: Identifier 'l' is never reassigned; use 'const' instead of 'let'.
src/input/ux-input.ts[53, 11]: Identifier 'attributeValue' is never reassigned; use 'const' instead of 'let'.
src/input/ux-input.ts[62, 11]: Identifier 'attributeValue' is never reassigned; use 'const' instead of 'let'.
src/input/ux-input.ts[102, 9]: Identifier 'blurEvent' is never reassigned; use 'const' instead of 'let'.
src/input/ux-input.ts[115, 9]: Identifier 'blurEvent' is never reassigned; use 'const' instead of 'let'.
src/textarea/ux-textarea.ts[52, 11]: Identifier 'attributeValue' is never reassigned; use 'const' instead of 'let'.
src/textarea/ux-textarea.ts[93, 9]: Identifier 'blurEvent' is never reassigned; use 'const' instead of 'let'.
src/textarea/ux-textarea.ts[106, 9]: Identifier 'blurEvent' is never reassigned; use 'const' instead of 'let'.
```

```
npm ERR! Windows_NT 10.0.14393
npm ERR! argv "C:\\Program Files\\nodejs\\node.exe" "C:\\Program Files\\nodejs\\node_modules\\npm\\bin\\npm-cli.js" "run" "lint"
npm ERR! node v7.7.4
npm ERR! npm  v4.1.2
npm ERR! code ELIFECYCLE
npm ERR! aurelia-ux@0.1.16 lint: `cross-env tslint --project tsconfig.json`
npm ERR! Exit status 2
npm ERR!
npm ERR! Failed at the aurelia-ux@0.1.16 lint script 'cross-env tslint --project tsconfig.json'.
npm ERR! Make sure you have the latest version of node.js and npm installed.
npm ERR! If you do, this is most likely a problem with the aurelia-ux package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     cross-env tslint --project tsconfig.json
npm ERR! You can get information on how to open an issue for this project with:
npm ERR!     npm bugs aurelia-ux
npm ERR! Or if that isn't available, you can get their info via:
npm ERR!     npm owner ls aurelia-ux
npm ERR! There is likely additional logging output above.

npm ERR! Please include the following file with any support request:
npm ERR!     E:\Development\Aurelia\ux\npm-debug.log
```

**Expected/desired behavior:**

For `npm run lint` to pass

### PR DESCRIPTION
**Pull request by ZHollingshead**
Most changes were of the following variety:

* change `let` to `const`
* change casting with `<>` to `as`
* change `Object` param type to `object`

fixes #43

### CODE REVIEW

File: tslint.json [hunk 1]
```diff
@@ -8,7 +8,12 @@
...
     "no-string-literal": false,
-    "trailing-comma": false,
-    "member-ordering": ["fields-first"]
+    "trailing-comma": [false],
+    "member-ordering": ["fields-first"],
+    "array-type": ["array-simple"],
+    "arrow-parens": false,
+    "max-classes-per-file": [false],
+    "prefer-for-of": false,
+    "ban-types": false
```
Comment (ZHollingshead) [inline tslint.json]: I am not sure we needed this rule to be enabled, and it was turned on after the tslint update. I turned it off for now as I did not see a good way to refactor many of the constructors that were using `Function` as a type.

File: tslint.json [hunk 2]
```diff
@@ -8,7 +8,12 @@
     "interface-name": [true, "never-prefix"],
     "no-shadowed-variable": false,
     "no-string-literal": false,
-    "trailing-comma": false,
-    "member-ordering": ["fields-first"]
+    "trailing-comma": [false],
+    "member-ordering": ["fields-first"],
+    "array-type": ["array-simple"],
```
Comment (ZHollingshead) [inline tslint.json]: Added new rules from the aurelia/validation library.

**Comment by khuongduybui** [comment]
Don't have permissions to approve, but this LGTM

File: tslint.json (hunk 1 above)
Comment (jdanyow) [inline tslint.json]: I think you can use `fn :() => void` in place of `fn: Function`. Of course, the more specific we can be about the function signature, the better. `Function` kind of acts like `any`- I think that's one reason it's "banned" by default.

Comment (ZHollingshead) [inline tslint.json]: This works somewhat. 

`() => void` will not work but `() => any` will.

The exception to this is dynamic-styles.ts

``` typescript
export interface StyleModule {
  [x: string]: () => any;
}

export function createDynamicStyleModule(styleUrl: string): StyleModule {
  @styles()
  @useStyles(styleUrl)
  class DynamicTheme {}

  return {
    [getNextDynamicThemeId()]: DynamicTheme
  };
}
```

Which causes the error.

```
src/styles/dynamic-styles.ts(18,3): error TS2322: Type '{ [x: string]: typeof DynamicTheme; }' is not assignable to type 'StyleModule'.
[4]   Index signatures are incompatible.
[4]     Type 'typeof DynamicTheme' is not assignable to type '() => any'.
[4]       Type 'typeof DynamicTheme' provides no match for the signature '(): any'
```

I am still fairly new to Typescript overall, so I am not entirely sure how to handle that. I have tried to set it as `() => DynamicTheme` but that did not work as expected.

**Pull request merged**